src/lib/slew_rate/SlewRate.hpp

```diff
@@ -0,0 +1,101 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file SlewRate.hpp
+ *
+ * Library limit the rate of change of a value with a maximum slew rate.
+ *
+ * @author Matthias Grob <maetugr@gmail.com>
+ */
+
+#pragma once
+
+#include <mathlib/mathlib.h>
+#include <matrix/math.hpp>
+
+template<typename Type>
+class SlewRate
+{
+public:
+	SlewRate() = default;
+	~SlewRate() = default;
+
+	/**
+	 * Set maximum rate of change for the value
+	 * @param slew_rate maximum rate of change
+	 */
+	void setSlewRate(const Type slew_rate) { _slew_rate = slew_rate; }
+
+	/**
+	 * Set value ignoring slew rate for initialization purpose
+	 * @param value new applied value
+	 */
+	void setForcedValue(const Type value) { _value = value; }
+
+	/**
+	 * Update slewrate
+	 * @param new_value desired new value
+	 * @param deltatime time in seconds since last update
+	 * @return actual value that complies with the slew rate
+	 */
+	Type update(const Type new_value, const float deltatime)
+	{
+		// Limit the rate of change of the value
+		const Type dvalue_desired = new_value - _value;
+		const Type dvalue_max = _slew_rate * deltatime;
+		const Type dvalue = constrain(dvalue_desired, -dvalue_max, dvalue_max);
+		_value += dvalue;
+		return _value;
+	}
+
+private:
+	Type _slew_rate{}; ///< maximum rate of change for the value
+	Type _value{}; ///< state to keep last value of the slew rate
+
+	float constrain(float value, float min, float max)
+	{
+		return math::constrain(value, min, max);
```

Comment:
Why does this exist?

Reply:
To have one implementation for float and one for Vector3f. I'll make it supported by the matrix library and this local implementation can go away.